api/internal/localizer/locloader.go

```diff
@@ -0,0 +1,351 @@
+		return nil, "", errors.WrapPrefixf(err, "invalid localize destination '%s'", newDirArg)
+	}
+
+	var realScope string
+	if repo, remote := ldr.Repo(); remote {
+		realScope = repo
+	} else {
+		realScope = scope.String()
+	}
+	toDst := pathSuffix(realScope, ldr.Root())
+
+	return &LocLoader{
+		fSys: fSys,
+		args: &locArgs{
+			scope:  scope,
+			newDir: newDir,
+		},
+		ldr:   ldr,
+		local: scope != "",
+		dst:   newDir.Join(toDst),
+	}, newDir, nil
+}
+
+// validateScope returns the scope given localize arguments and targetLdr that corresponds to targetArg
+func validateScope(scopeArg string, targetArg string, targetLdr ifc.Loader, fSys filesys.FileSystem) (filesys.ConfirmedDir, error) {
+	var scope filesys.ConfirmedDir
+
+	errBadScopeForTarget := fmt.Sprintf("invalid localize scope '%s' for target '%s'", scopeArg, targetArg)
+	switch _, remote := targetLdr.Repo(); {
+	case remote:
+		if scopeArg != "" {
+			return "", errors.WrapPrefixf(
+				errors.Errorf("cannot specify scope for remote localize target"),
+				errBadScopeForTarget)
+		}
+	case scopeArg == "":
+		scope = filesys.ConfirmedDir(targetLdr.Root())
+	default:
+		var err error
+		scope, err = filesys.ConfirmDir(fSys, scopeArg)
+		if err != nil {
+			return "", errors.WrapPrefixf(err, "unable to establish localize scope")
+		}
+		if !filesys.ConfirmedDir(targetLdr.Root()).HasPrefix(scope) {
+			return scope, errors.WrapPrefixf(errRootOutsideScope, errBadScopeForTarget)
+		}
+	}
+	return scope, nil
+}
+
+// parseLocRootURL accepts a remote root rootURL and returns its RepoSpec if it satisfies localize requirements;
+// otherwise, an error.
+func parseLocRootURL(rootURL string) (*git.RepoSpec, error) {
+	spec, err := git.NewRepoSpecFromURL(rootURL)
+	if err != nil {
+		log.Fatalf(errors.WrapPrefixf(
+			err,
+			"%s: cannot parse validated remote root '%s'", fatalPrefix, rootURL).Error())
+	}
+	if spec.Ref == "" {
+		return nil, errors.WrapPrefixf(
+			errors.Errorf("localize remote root missing ref query string parameter"),
+			"invalid localize remote root '%s'", rootURL)
+	}
+	return spec, nil
+}
+
+// urlBase is the url equivalent of filepath.Base
+func urlBase(url string) string {
+	cleaned := strings.TrimRight(url, urlSeparator)
+	i := strings.LastIndex(cleaned, urlSeparator)
+	if i < 0 {
+		return cleaned
+	}
+	return cleaned[i+1:]
+}
+
+// defaultNewDir calculates the default localize destination directory name from targetLdr at the localize target
+// and spec of target, which is nil if target is local
+func defaultNewDir(targetLdr ifc.Loader, spec *git.RepoSpec) string {
+	targetDir := filepath.Base(targetLdr.Root())
+	switch repo, remote := targetLdr.Repo(); {
+	case remote:
+		if repo == targetLdr.Root() {
+			targetDir = urlBase(spec.OrgRepo)
+		}
+		return strings.Join(
+			[]string{dstPrefix, targetDir, strings.ReplaceAll(spec.Ref, urlSeparator, replacement)},
+			replacement)
+	case targetDir == string(filepath.Separator):
+		return dstPrefix
+	default:
```

Comment:
leave comments explaining what each of these cases means

Reply:
Done.